TerrariaServerAPI/TerrariaApi.Server/Hooking/NetHooks.cs

```diff
@@ -21,132 +21,141 @@ public static void AttachTo(HookManager hookManager)
-			Hooks.Net.BeforeBroadcastChatMessage = OnBroadcastChatMessage;
-			Hooks.Player.NameCollision = OnNameCollision;
+			On.Terraria.NetMessage.greetPlayer += OnGreetPlayer;
+			On.Terraria.Netplay.OnConnectionAccepted += OnConnectionAccepted;
+			On.Terraria.Chat.ChatHelper.BroadcastChatMessage += OnBroadcastChatMessage;
+			On.Terraria.Net.NetManager.SendData += OnSendNetData;
+
+			Hooks.NetMessage.SendData += OnSendData;
+			Hooks.NetMessage.SendBytes += OnSendBytes;
+			Hooks.MessageBuffer.GetData += OnReceiveData;
+			Hooks.MessageBuffer.NameCollision += OnNameCollision;
 		}
 
-		private static HookResult OnBroadcastChatMessage(NetworkText text, ref Color color, ref int ignorePlayer)
+		static void OnBroadcastChatMessage(On.Terraria.Chat.ChatHelper.orig_BroadcastChatMessage orig, NetworkText text, Color color, int excludedPlayer)
 		{
 			float r = color.R, g = color.G, b = color.B;
 
 			var cancel = _hookManager.InvokeServerBroadcast(ref text, ref r, ref g, ref b);
 
-			color.R = (byte)r;
-			color.G = (byte)g;
-			color.B = (byte)b;
+			if (!cancel)
+			{
+				color.R = (byte)r;
+				color.G = (byte)g;
+				color.B = (byte)b;
 
-			return cancel ? HookResult.Cancel : HookResult.Continue;
+				orig(text, color, excludedPlayer);
+			}
 		}
 
-		static HookResult OnSendData(
-				ref int bufferId,
-				ref int msgType,
-				ref int remoteClient,
-				ref int ignoreClient,
-				ref Terraria.Localization.NetworkText text,
-				ref int number,
-				ref float number2,
-				ref float number3,
-				ref float number4,
-				ref int number5,
-				ref int number6,
-				ref int number7
-			)
+		static void OnSendData(object sender, Hooks.NetMessage.SendDataEventArgs e)
 		{
-			if (_hookManager.InvokeNetSendData
-			(
-				ref msgType,
-				ref remoteClient,
-				ref ignoreClient,
-				ref text,
-				ref number,
-				ref number2,
-				ref number3,
-				ref number4,
-				ref number5,
-				ref number6,
-				ref number7
-			))
+			if (e.Event == HookEvent.Before)
 			{
-				return HookResult.Cancel;
+				var msgType = e.MsgType;
+				var remoteClient = e.RemoteClient;
+				var ignoreClient = e.IgnoreClient;
+				var text = e.Text;
+				var number = e.Number;
+				var number2 = e.Number2;
+				var number3 = e.Number3;
+				var number4 = e.Number4;
+				var number5 = e.Number5;
+				var number6 = e.Number6;
+				var number7 = e.Number7;
+				if (_hookManager.InvokeNetSendData
+				(
+					ref msgType,
+					ref remoteClient,
+					ref ignoreClient,
+					ref text,
+					ref number,
+					ref number2,
+					ref number3,
+					ref number4,
+					ref number5,
+					ref number6,
+					ref number7
+				))
+				{
+					e.Result = HookResult.Cancel;
+				}
+
+				e.MsgType = msgType;
+				e.RemoteClient = remoteClient;
+				e.IgnoreClient = ignoreClient;
+				e.Text = text;
+				e.Number = number;
+				e.Number2 = number2;
+				e.Number3 = number3;
+				e.Number4 = number4;
+				e.Number5 = number5;
+				e.Number6 = number6;
+				e.Number7 = number7;
```

Comment:
Previous functionality would pass `HookResult.Continue` but this doesn't

Reply:
this is a monomod hook, not a otapi hook. instead it just doesnt call the 'orig' method, like the HookResult would have done anyway (just via injected IL).